### PR TITLE
add error messages to assert_eq()

### PR DESCRIPTION
**Pull request by jameslamb**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

## Changes in this PR

This PR proposes adding more informative error messages in `dask.array.utils.assert_eq()`.

## How this improves `dask`

We make heavy use of `assert_eq()` in [LightGBM's Dask tests](https://github.com/microsoft/LightGBM/blob/8593f8503cb004fe120ee53420bd272bde001f55/tests/python_package_test/test_dask.py). I've had the following experience a few times:

1. make changes
2. run `pytest`
3. tests fail. To figure out how they failed, scroll up to find the `assert_eq()` information dumped into the console, then try to understand the line in the code that's pointed to.

image

As of this PR, `assert_eq()` would raise an `AssertionError` with enough information to understand **how** the inputs `a` and `b` differ. For example:

image

> AssertionError: a and b have different shapes (a: (100,), b: (100, 2))

## Notes for Reviewers

I wasn't sure where it would be appropriate to add tests on these changes. `tests/test_array_utils.py` doesn't currently contain tests on `assert_eq()` directly. If you're open to the change proposed in this PR, please let me know if you'd like me to add some tests on the specific messages as well.

Thanks for your time and consideration.

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the PR @jameslamb! Adding informative error messages here is a nice improvement

File: dask/array/utils.py [hunk 1]
```diff
@@ -270,23 +270,50 @@ def assert_eq(a, b, check_shape=True, check_graph=True, check_meta=True, **kwarg
...
+                msg = (
+                    f"compute()-ing 'b' changes its number of dimensions "
+                    f"(before: {b_original._meta.ndim}, after: {b.ndim})"
+                )
+                assert b_original._meta.ndim == b.ndim, msg
                 if b_meta is not None:
-                    assert type(b_original._meta) == type(b_meta)
+                    msg = (
+                        f"compute()-ing 'b' changes its type "
+                        f"(before: {type(a_original._meta)}, after: {type(a_meta)})"
```
Comment (jrbourbeau) [inline dask/array/utils.py:306]:
```suggestion
                        f"(before: {type(b_original._meta)}, after: {type(b_meta)})"
```

File: dask/array/utils.py [hunk 2]
```diff
@@ -270,23 +270,50 @@ def assert_eq(a, b, check_shape=True, check_graph=True, check_meta=True, **kwarg
...
+                    msg = (
+                        f"compute()-ing 'a' changes its type "
+                        f"(before: {type(a_original._meta)}, after: {type(a_meta)})"
+                    )
+                    assert type(a_original._meta) == type(a_meta), msg
                     if not (np.isscalar(a_meta) or np.isscalar(a_computed)):
-                        assert type(a_meta) == type(a_computed)
+                        msg = (
+                            f"compute()-ing 'a' results in a different type than implied by its metadata "
+                            f"(meta: {type(b_meta)}, computed: {type(b_computed)})"
```
Comment (jameslamb) [inline dask/array/utils.py:294]:
```suggestion
                            f"(meta: {type(a_meta)}, computed: {type(a_computed)})"
```

Comment (jameslamb) [reply dask/array/utils.py:294]: bad copy-paste, sorry :grimacing:

**Review by jrbourbeau — APPROVED** [review]
Thanks @jameslamb! This is in

**Comment by jrbourbeau** [comment]
FWIW there's an open issue for adding tests around out `assert_eq` testing utilities. Any additions on that front are certainly welcome, though no obligation

**Comment by jrbourbeau** [comment]
Linking to the issue https://github.com/dask/dask/issues/2682

**Comment by jameslamb** [comment]
thanks, I'll take a look!

**Pull request merged**